### PR TITLE
Creating a data model for a Blog App

### PR DESCRIPTION
**Pull request by MarcoDiaz2000**
## Description
In this project, I used migration files to build your schema.

blog_app_erd

### General requirements

- Make sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
- Make sure that you used correct flow [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
- Make sure that you documented your work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).

### Project requirements
Be sure to reference the [ERD diagram](../images/blog_app_erd.png) from the sneak peek if you have any questions about table or column names or data relationships.

- Create a new app. Give it a meaningful name.
- Initialize your project with Git.
- Make sure that your project has Postgres database set up.
- Build your project schema.
    - Create and run the necessary migration files.
    - Table and column names should match the ERD diagram.
        - note: photo for users table should be a link to an image
    - Foreign keys should be included.
        - Pay attention that in `posts` table the column holding the foreign key to the `users` table should be named `author_id`
    - All columns that are foreign keys should have a corresponding index.

### CODE REVIEW

File: db/migrate/20230815130251_create_posts.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class CreatePosts < ActiveRecord::Migration[7.0]
+  def change
+    create_table :posts do |t|
+      t.integer :author_id
```
Comment (bienvenuushindi) [inline db/migrate/20230815130251_create_posts.rb]: :red_circle: [REQUIRED] Please make sure you add the foreign key constraint that ensures the author_id in the posts table references a valid id in the authors table. This helps maintain data integrity and consistency in your database.

Here's an example of how you might modify your code to include a foreign key constraint:

```
def change
  create_table :posts do |t|
    t.references :author, foreign_key: true # This adds the foreign key constraint
   # other columns
  end
end

```

Importance of Using Foreign Key Constraints:

1. Data Integrity: Foreign key constraints help maintain the integrity of your data by preventing invalid relationships. They ensure that data in the referencing table (posts) matches data in the referenced table (authors), preventing orphaned or mismatched records.

2. Database Integrity: Using foreign keys enhances the integrity of your database as a whole, preventing scenarios where data inconsistencies can occur due to missing or incorrect relationships.

3. Data Accuracy: By enforcing valid relationships, foreign keys prevent accidental data corruption, improving the accuracy of your application's data.

4. Ease of Maintenance: Foreign key constraints can make database maintenance and updates more straightforward. They help you avoid errors when updating or deleting records by automatically ensuring data consistency.

5. Performance: While foreign keys might have a minor performance impact due to additional checks, they can also improve query performance by enabling the database to optimize related queries.

6. Documentation: Foreign keys serve as documentation for your database schema, making it clear which tables are related and how they're linked.

In summary, using foreign key constraints is a best practice for maintaining data integrity and consistency in relational databases. They help prevent data anomalies, enhance data accuracy, and simplify maintenance tasks. When designing a database schema, it's highly recommended to leverage foreign key constraints wherever applicable.

File: db/migrate/20230815130345_create_comments.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class CreateComments < ActiveRecord::Migration[7.0]
+  def change
+    create_table :comments do |t|
+      t.integer :author_id
+      t.integer :post_id
```
Comment (bienvenuushindi) [inline db/migrate/20230815130345_create_comments.rb]: :red_circle: [REQUIRED] Similar to the previous example, you can add foreign key constraints to ensure that the author_id and post_id fields in the comments table reference valid id values in their respective tables (authors and posts).

Here's how you can do that:

```
create_table :comments do |t|
    t.references :author, foreign_key: true # Add foreign key constraint for author
    t.references :post, foreign_key: true   # Add foreign key constraint for post
    t.text :text

    t.timestamps
  end
```
In this code, the lines t.references :author, foreign_key: true and t.references :post, foreign_key: true add foreign key constraints to the author_id and post_id fields, respectively. These constraints ensure that the values in these fields reference valid records in the authors and posts tables, maintaining the integrity of your data.

File: db/migrate/20230815131638_create_likes.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class CreateLikes < ActiveRecord::Migration[7.0]
+  def change
+    create_table :likes do |t|
+      t.integer :author_id
+      t.integer :post_id
```
Comment (bienvenuushindi) [inline db/migrate/20230815131638_create_likes.rb]: :red_circle: [REQUIRED] Kindly modify the code for the likes table to include foreign key constraints. This will ensure that the author_id and post_id fields reference valid records in their respective tables (authors and posts), just like in the previous examples.

Here's how you can add foreign key constraints to the likes table:
 ```
create_table :likes do |t|
    t.references :author, foreign_key: true # Add foreign key constraint for author
    t.references :post, foreign_key: true   # Add foreign key constraint for post

    t.timestamps
  end
```


In this code, the lines t.references :author, foreign_key: true and t.references :post, foreign_key: true add foreign key constraints to the author_id and post_id fields, respectively. These constraints ensure that the values in these fields reference valid records in the authors and posts tables, maintaining data integrity.

By using foreign key constraints in the likes table, you're applying the same principles of data integrity and accuracy as discussed earlier. It prevents data inconsistencies and ensures that the relationships between records are valid, making your database more reliable and maintainable.

**Review by bienvenuushindi — CHANGES_REQUESTED** [review]
Hi @MarcoDiaz2000  , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights
May I highlight the following:

- The repo looks professional.:heavy_check_mark:
- There are no linter errors:heavy_check_mark:
- You have descriptive PR:heavy_check_mark:
- Your project looks good:heavy_check_mark:

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Check the comments under the review._

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: db/migrate/20230815130251_create_posts.rb (hunk 1 above)
Comment (MarcoDiaz2000) [reply db/migrate/20230815130251_create_posts.rb]: Thanks

**Review by KDlamini — APPROVED** [review]
### Hi @MarcoDiaz2000 🙋‍♀️,  

Excellent work! 💯 
You have done a great job implementing all the project requirements for creating a data model for the Blog App. 
Your feature is complete! There is nothing else to say other than... it's time to merge it. :shipit: 
Congratulations!!!🎊 🥳 

## Highlights 🌞 

✔️  Good PR title and description
✔️  Well-documented and professional README
✔️  Migration files for users, posts, comments, and likes are implemented
✔️  Indexes are included in relevant tables

# Status: Approved ✔️ 

giphyapproved
 
Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me @KDlamini  in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by MarcoDiaz2000** [comment]
> ### Hi @MarcoDiaz2000 🙋‍♀️,
> Excellent work! 💯 You have done a great job implementing all the project requirements for creating a data model for the Blog App. Your feature is complete! There is nothing else to say other than... it's time to merge it. :shipit: Congratulations!!!🎊 🥳
> 
> ## Highlights 🌞
> ✔️ Good PR title and description ✔️ Well-documented and professional README ✔️ Migration files for users, posts, comments, and likes are implemented ✔️ Indexes are included in relevant tables
> 
> # Status: Approved ✔️
> giphyapproved [ giphyapproved ](https://user-images.githubusercontent.com/38088818/152676874-3b2edc2c-a46f-4da6-b253-be859d2a2c65.gif) [ ](https://user-images.githubusercontent.com/38088818/152676874-3b2edc2c-a46f-4da6-b253-be859d2a2c65.gif)
> 
> Cheers 🥂 and Happy coding!!! 👯 Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Remember to tag me @KDlamini in your question so I can receive the notification.
> 
> _As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Thanks
